### ISSUE TITLE
GaussianFitter is very fragile

### ISSUE THREAD

**Issue by joefowler** [usnistgov/mass#58]
**[Original report](https://bitbucket.org/joe_fowler/mass/issues/58) by Galen O'Neil (Bitbucket: [oneilg](https://bitbucket.org/oneilg), GitHub: [oneilg](https://github.com/oneilg)).**

----------------------------------------

For example

```
fitter = mass.calibration.GaussianFitter()
bin_centers = np.arange(885, 940)
trueparams = [10,930,100,0,0,0.11,28]
guessparams = [10, 920, 100, 0,0,0.17,14]
fakedata = fitter.fitfunc(trueparams,bin_centers)
fitter.fit(fakedata, bin_centers, params=guessparams)
fitter.fit(fakedata, bin_centers, vary_bg_slope=True, vary_bg=True, vary_tail=True)
```

Gives me a singular matrix error. I feel like it is way more fragile than the KAlpha and KBeta fitters.

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

What worries me most is that 

```
#!python

fitter.fit(fakedata, bin_centers, params=trueparams)

```
also gives a singular matrix error. Even if you get the parameters correct, you still have problems. Hmm.

**Comment by joefowler** [comment]
**Original comment by Galen O'Neil (Bitbucket: [oneilg](https://bitbucket.org/oneilg), GitHub: [oneilg](https://github.com/oneilg)).**

----------------------------------------

I found that ALL fitters fail when `background=0` in the guess params. In commit b85abce I added tests to test_fits with bg = 0, they currently all fail. I'm planing to merge that into develop eventually, but for now it's just in hdf5_only, which is why I gave you the specific commit. I don't know how to fix it, but at least we have a more specific source of error.

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

Hey, this is definitely a useful cause. I wonder if it's a weird interaction in the way that "bounded parameters" are implemented. Background is a parameter that is necessarily non-negative, which under the hood we accomplish by using a hidden function to transform the visible parameter into an internal one that can run from -inf to +inf. I wonder if there's something tricky going on when you try to set the parameter to the exact boundary value.

Would you try your tests with `background=1e-6` or something? This has two purposes: first, if successful, it's a nice workaround for you so you can keep doing whatever you're doing. Second, it would be useful information for me regarding this bug (so let me know whether this works).

**Comment by joefowler** [comment]
**Original comment by Galen O'Neil (Bitbucket: [oneilg](https://bitbucket.org/oneilg), GitHub: [oneilg](https://github.com/oneilg)).**

----------------------------------------

I tried (before you sent this response) and found that it started crashing at about 1e-21, I probably took steps about about 5 orders of magnitude.

I changed some code we were using to estimate the background to use float division instead of integer division, which is effectively the implementing your suggestion. It seems to be working. But it used to work with the integer division, suggesting this is a regression.

Also I feel like now that we’re actually starting to implement tests that it feels like a much more manageable and rewarding task than I had expected. Once you get the github response, I’ll try to get some automated testing set up.

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

I can definitely see how this would be a problem that makes an (unwanted) distinction between true zero and near-zero. Your tests say yes. But I cannot think of how this could be a regression. Well, it apparently is, whether I can think of a good mechanism or not.

I have not heard back from Aaron Fein about github membership. I'll write to ask.

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

Thanks for the test functions in commit b85abce. I am using them now. It's definitely a problem with setting the initial value of a parameter right at a boundary (or within epsilon of it). I am working out what's the right solution.

**Comment by joefowler** [comment]
**Original comment by Joseph Fowler (Bitbucket: [joe_fowler](https://bitbucket.org/joe_fowler), ).**

----------------------------------------

Make MaximumLikelihoodHistogramFitter more robust.

Specifically, (1) force bounded parameters into the bounded range, and
(2) "kick" bounded parameters away from their bounds by up to one epsilon,
where the latter is the step size used when taking numerical gradients.

Fixes #58.
